### PR TITLE
Docs - CUDA suport for PL/Container

### PR DESCRIPTION
**Pull request by dyozie**
Add requirements and instructions for CUDA functions in PL/Container Python.

HTML of this PR available at https://docs-staging.vmware.com/en/VMware-Tanzu-Greenplum/cuda/greenplum-database/GUID-analytics-pl_container_using.html#developing-cuda-api-functions-with-plcontainer-9

### CODE REVIEW

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 1]
```diff
@@ -290,7 +288,179 @@ Keep in mind that UDFs that you created for Python 2 may not run in PL/Container
...
+
+Beginning with version 2.2, PL/Container supports developing Compute Unified Device Architecture (CUDA) API functions that utilize NVIDIA GPU hardware. This is accomplished by using the NVIDIA Container Toolkit `nvidia-docker` image and the `pycuda` python library. This procedure explains how to set up PL/Container for developing these functions.
+
+#### Prerequisites
+
+To develop CUDA functions with PL/Container you require:
+- A Docker installation having Docker engine version v19.03 or newer 
+- PL/Container version 2.2.0 or newer 
+- At least one NVIDIA GPU with the required GPU driver installed on your host
+- An installation of the
```
Comment (Sasasu) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:301]: ( I think this need to be removed.

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 2]
```diff
@@ -290,7 +288,179 @@ Keep in mind that UDFs that you created for Python 2 may not run in PL/Container
...
+#### Install and Customize the PL/Container Image
+
+1. Download the `plcontainer-python3-image-2.2.0-gp6.tar.gz` file from the **Greenplum Procedural Languages** section on [Tanzu Network](https://network.pivotal.io/products/vmware-tanzu-greenplum).
+
+2. Load the downloaded PL/Container image into Docker:
+    ```
+    $ docker image load < plcontainer-python3-image-2.2.0-gp6.tar.gz
+    ```
+
+3. Install the `python3-pycuda` software to your operating system. For example:
```
Comment (Sasasu) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:330]: our user needs to install two software. 1. the cuda runtime 2. the pycuda library

the Dockerfile is an example of how to install these two software, not the fourth step.

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 3]
```diff
@@ -290,7 +288,179 @@ Keep in mind that UDFs that you created for Python 2 may not run in PL/Container
...
+        python3.7 -m pip --no-cache-dir install platformdirs==2.5.2 && \ 
+        python3.7 -m pip --no-cache-dir install pytools==2022.1.2 && \ 
+        python3.7 -m pip --no-cache-dir install pycuda==2021.1 
+    ```
+
+5. Build the a customized container using your Dockerfile:
+    ```
+    $ docker build . -t localhost/plcontainer_python3_cuda_shared:latest
+    ```
+    **Note:** The remaining instructions use the example image tag `host/plcontainer_python3_cuda_shared:latest`. Substitute the actual tag name as needed.
```
Comment (Sasasu) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:371]:
```suggestion
    **Note:** The remaining instructions use the example image tag `localhost/plcontainer_python3_cuda_shared:latest`. Substitute the actual tag name as needed.
```

**Review by Sasasu — APPROVED** [review]

File: gpdb-doc/markdown/analytics/pl_container.html.md [hunk 4]
```diff
@@ -2,7 +2,7 @@
 title: PL/Container Language 
 ---
 
-Pl/Container enables users to run Greenplum procedural language functions inside a Docker container, to avoid security risks associated with running Python or R code on Greenplum segment hosts. This topic covers information about the architecture, installation, and setup of PL/Container:
+PL/Container enables users to run Greenplum procedural language functions inside a Docker container, to avoid security risks associated with running Python or R code on Greenplum segment hosts. For Python, PL/Container also enables you to use the Compute Unified Device Architecture (CUDA) API with Nvidia GPU hardware in your procedural language functions. This topic covers information about the architecture, installation, and setup of PL/Container:
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_container.html.md:5]: Nvidia -> NVIDIA (or vice versa) so refs are consistent throughout the pages?   most of the refs below are all caps.

File: gpdb-doc/markdown/analytics/pl_container.html.md [hunk 5]
```diff
@@ -70,13 +70,13 @@ The following sections describe these tasks in detail.
...
 -   For PL/Container 3 Beta use Greenplum Database 6.1 or later on CentOS 7.x \(or later\), RHEL 7.x \(or later\), or Ubuntu 18.04.
--   The minimum Linux OS kernel version supported is 3.10. To verfiy your kernel release use:
+-   The minimum Linux OS kernel version supported is 3.10. To verify your kernel release use:
 
     ```
     $ uname -r
     ```
 
--   The minimum Docker versions on all hosts needs to be Docker 19.03.
+-   The minimum supporteed Docker versions on all hosts is Docker 19.03.
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_container.html.md:79]: supporteed-> supported

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 6]
```diff
@@ -290,7 +288,179 @@ Keep in mind that UDFs that you created for Python 2 may not run in PL/Container
...
+
+#### Prerequisites
+
+To develop CUDA functions with PL/Container you require:
+- A Docker installation having Docker engine version v19.03 or newer 
+- PL/Container version 2.2.0 or newer 
+- At least one NVIDIA GPU with the required GPU driver installed on your host
+- An installation of the 
+
+See the [Getting Started](https://github.com/NVIDIA/nvidia-docker) section of the NVIDIA Container Toolkit GitHub project for Information about installing the NVIDIA driver or Docker engine for your Linux distribution.
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:303]: Information -> information ?

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 7]
```diff
@@ -290,7 +288,179 @@ Keep in mind that UDFs that you created for Python 2 may not run in PL/Container
...
+
+Follow the [Installation Guide](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html) for the NVIDIA Container Toolkit GitHub project to install the `nvidia-docker` container.
+
+Verify that the Docker image can use your installed GPU(s) by running the command similar to:
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:307]: the command similar -> a command similar ?

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 8]
```diff
@@ -290,7 +288,179 @@ Keep in mind that UDFs that you created for Python 2 may not run in PL/Container
...
+7. Edit the image runtime to assign a GPU. The following example adds GPU device ID `0` as the GPU, and `gpadmin` as the designated role. Substitute either the GPU device ID or the device UUID that you recorded earlier:
+    ```
+    $ plcontainer runtime-edit
+    ```
+    ```
+        <runtime> 
+            <id>plc_python_cuda_shared</id> 
+            <image>localhost/plcontainer_python3_cuda_shared:latest</image> 
+            <command>/clientdir/py3client.sh</command> 
+        <setting roles="gpadmin" />
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:387]: you closed the setting tag on the same line.  do the tags/settings below this line need to be indented?

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 9]
```diff
@@ -290,7 +288,179 @@ Keep in mind that UDFs that you created for Python 2 may not run in PL/Container
...
+    ```
+
+#### Create and Run a Sample CUDA Function
+
+1. Connect to a Greenplum database where PL/Container is installed:
+    ```
+    $ psql -d mytest -h master_host -p 5432 -U `gpadmin`
+    ```
+
+2. Create a sample PL/Container function that uses the container you customized (`plc_python_cuda_shared` in this example):
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:402]: can you provide a description of what the function does?  or include comments in the function?

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 10]
```diff
@@ -290,7 +288,179 @@ Keep in mind that UDFs that you created for Python 2 may not run in PL/Container
...
+            block=(400,1,1)) 
+      
+    return [float(i) for i in (dest-a*b)] 
+    
+    $$ LANGUAGE plcontainer; 
+    ```
+
+3. Run the sample function and verify its output:
+    ```
+    $ with a as (select unnest(hello) as cuda from hello_cuda() as hello) select sum(cuda) from a;
```
Comment (lisakowen) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:439]: should we capitalize the keywords here and below (WITH, AS, FROM, SELECT)?

**Review by lisakowen — COMMENTED** [review]
just a few minor comments.

**Review by Sasasu — APPROVED** [review]

**Pull request merged**